Bio/SeqIO/AbiIO.py

```diff
@@ -33,23 +33,250 @@
 # each key is tag_name + tag_number
 # if a tag entry needs to be added, just add its key and its key
 # for the annotations dictionary as the value
+# dictionary for tags that require preprocessing before use in creating
+# seqrecords
 _EXTRACT = {
     'TUBE1': 'sample_well',
     'DySN1': 'dye',
     'GTyp1': 'polymer',
     'MODL1': 'machine_model',
 }
-# dictionary for tags that require preprocessing before use in creating
-# seqrecords
-_SPCTAGS = [
-    'PBAS2',    # base-called sequence
-    'PCON2',    # quality values of base-called sequence
-    'SMPL1',    # sample id inputted before sequencing run
-    'RUND1',    # run start date
-    'RUND2',    # run finish date
-    'RUNT1',    # run start time
-    'RUNT2',    # run finish time
-]
+
+
+# Complete data structure representing 98% of the API. The general section
+# represents the part of the API that's common to ALL instruments, whereas the
+# instrument specific sections are labelled as they are in the ABIF spec
+#
+# The 2%:
+#  - on some fields, there are repeated fields (e.g DYEN1-100), and those are not supported
+#  - OvrI1-N containing color data of some sort
+#
+# Keys don't seem to clash from machine to machine, so when we parse, we look
+# for ANY key, and store that in the raw ABIF data structure attached to the
+# annotations, with the assumption that anyone parsing the data can look up the
+# spec themself
+#
+# Key definitions are retained in case end users want "nice" labels pre-made
+# for them for all of the available fields.
+_INSTRUMENT_SPECIFIC_TAGS = {
+    'general': {
+        'APFN2': 'Sequencing Analysis parameters file name',
+        'APXV1': 'Analysis Protocol XML schema version',
+        'APrN1': 'Analysis Protocol settings name',
+        'APrV1': 'Analysis Protocol settings version',
+        'APrX1': 'Analysis Protocol XML string',
+        'CMNT1': 'Sample Comment',
+        'CTID1': 'Container Identifier, a.k.a. plate barcode',
+        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
+        'CTTL1': 'Comment Title',
+        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
+        'DATA1': 'Channel 1 raw data',
+        'DATA1': 'Channel 1 raw data',
```

Comment:
I was referring to checking the instrument ID and then adding all the keys
that have different keys on different instruments (
http://www6.appliedbiosystems.com/support/software_community/ABIF_File_Format.pdf).
However, this may require revision of the SeqRecord file. As for the POS
tag, I was generically referring to the peak position data, which has
different tags for different instruments. By combining the peak position
key with the DATA keys, one can create X-Y plots of the sequence results
and perform various statistics on the sequencing data (relative peak
heights, area under the curve, etc).

Thanks,
David

On Thu, Feb 12, 2015 at 9:21 AM, Eric Rasche <notifications@github.com>
wrote:

> In Bio/SeqIO/AbiIO.py
> <https://github.com/biopython/biopython/pull/473#discussion_r24583286>:
>
> > +# Key definitions are retained in case end users want "nice" labels pre-made
> > +# for them for all of the available fields.
> > +_INSTRUMENT_SPECIFIC_TAGS = {
> > +    'general': {
> > +        'APFN2': 'Sequencing Analysis parameters file name',
> > +        'APXV1': 'Analysis Protocol XML schema version',
> > +        'APrN1': 'Analysis Protocol settings name',
> > +        'APrV1': 'Analysis Protocol settings version',
> > +        'APrX1': 'Analysis Protocol XML string',
> > +        'CMNT1': 'Sample Comment',
> > +        'CTID1': 'Container Identifier, a.k.a. plate barcode',
> > +        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
> > +        'CTTL1': 'Comment Title',
> > +        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
> > +        'DATA1': 'Channel 1 raw data',
> > +        'DATA1': 'Channel 1 raw data',
>
> @bow <https://github.com/bow> nice catch, thanks!
>
> @DavidBulger <https://github.com/DavidBulger> I didn't see the POS tag in
> the spec, that sounds quite useful. I have a couple dozen test files, but
> they're pretty assuredly all from the same instrument, so I'd be very
> interested in having some more test data.
> I had avoided implementing any type of instrument detection for the very
> reason of not having any test data to validate my answers. Additionally I
> was concerned that testing for the presence/abscence of some tags as an
> instrument detection method would be error-prone. For example, see
> abi_prism_3100/3100-Avant which has a few different keys from the general
> set, but all of the keys there are optional.
>
> "Additional useful tags available...if we...check which ABI version is
> being used and then add tags specific for that version"
>
> Currently the abif_raw data structure comes back with any found keys from
> any instrument, see L427
> <https://github.com/biopython/biopython/pull/473/files#diff-9bba5ad30827093582367314de9a01ebR427>,
> I just merge all instrument keys available as there aren't
> instrument-specific differences in key definitions. Or did you mean
> something else by that?
>
> Thanks for the feedback, all!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biopython/biopython/pull/473/files#r24583286>.
>